### PR TITLE
Add a time-based decay to the linear allocation model.

### PR DESCRIPTION
**Pull request by PeterSolMS**
Real world scenario that motivated this allocated a huge amount of data once a day, leading to high gen 0 and gen 1 budgets (several GB).

After this, there was relatively little activity, cause gen 1 budget to take several hours to normalize.

The new logic discounts the previous desired allocation over a period of 5 minutes.

### CODE REVIEW

**Comment by PeterSolMS** [comment]
This addresses issue https://github.com/dotnet/runtime/issues/52592.

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -36767,6 +36765,9 @@ size_t gc_heap::desired_new_allocation (dynamic_data* dd,
         float     f = 0;
         size_t    max_size = dd_max_size (dd);
         size_t    new_allocation = 0;
+        float     time_since_previous_collection_secs = (dd_time_clock (dd) - dd_previous_time_clock (dd))*1e-6f;
+
+
```
Comment (mangod9) [inline src/coreclr/gc/gc.cpp:36770]: Nit: remove one empty line?

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -36726,21 +36728,17 @@ float gc_heap::surv_to_growth (float cst, float limit, float max_limit)
 //not be correct (collection happened too soon). Correct with a linear estimation based on the previous
 //value of the budget
 static size_t linear_allocation_model (float allocation_fraction, size_t new_allocation,
-                                       size_t previous_desired_allocation, size_t collection_count)
+                                       size_t previous_desired_allocation, float time_since_previous_collection_secs)
 {
     if ((allocation_fraction < 0.95) && (allocation_fraction > 0.0))
     {
+        const float decay_time = 5*60.0f; // previous desired allocation expires over 5 minutes
+        float decay_factor = (decay_time <= time_since_previous_collection_secs) ?
```
Comment (mangod9) [inline src/coreclr/gc/gc.cpp:36736]: will we build a test scenario in GCPerfSim to validate that the decay has the desired effect?

**Comment by Maoni0** [comment]
this seems to be along the same lines of what we are doing for WKS GC in `generation_to_condemn` where we check against `dd_time_clock_interval` so I'm wondering if we should take advantage of that, ie, use `dd_time_clock_interval` (but we should specify different values for Server GC - I would suggest much bigger values there) as decay_time so it's generation aware. 

also the way the change is now is modifying the existing behavior, it seems like it would be better to not modify the existing behavior when a GC happens soon enough, and start decaying when the elapsed time between this and last GC (of that generation) is longer than decay_time (and the longer it has been the less effect it should have). what do you think?

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:36770]: Looking at the code, I think I should remove both of the empty lines - no reason to separate the declarations here, I think.

File: src/coreclr/gc/gc.cpp (hunk 2 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:36736]: I will try, but it may be a bit involved.

**Comment by PeterSolMS** [comment]
I thought about making the decay_time generation aware along the lines you suggested, but didn't see a good reason to do it. Gen 1 is the only generation where we collect routinely without having exhausted the budget.

Regarding your remark about modifying existing behavior, I did this on purpose so a case where gen 1 GCs happen often enough, but very early (i.e. only exhausting a tiny fraction of the gen 1 budget), wouldn't lead to us taking a long time to ramp down the gen 1 budget. As it happens, a decay time of 5 minutes is short enough to fix the test case (which does about 6 gen 1 GCs in the first hour after loading the big chunk of data), but it seems unwise to rely on this - with more load, of course the interval between gen 1 GCs may drop below the decay_time.

**Comment by Maoni0** [comment]
> I thought about making the decay_time generation aware along the lines you suggested, but didn't see a good reason to do it. Gen 1 is the only generation where we collect routinely without having exhausted the budget.

in high memory load situation we could also collect gen2 often without having exhausted the budget but you could argue that the next gen2 will also happen without having exhausted budget due to the same reason and also gen2's surv rate is usually very high anyway. 

> Regarding your remark about modifying existing behavior, I did this on purpose so a case where gen 1 GCs happen often enough, but very early (i.e. only exhausting a tiny fraction of the gen 1 budget), wouldn't lead to us taking a long time to ramp down the gen 1 budget. As it happens, a decay time of 5 minutes is short enough to fix the test case (which does about 6 gen 1 GCs in the first hour after loading the big chunk of data), but it seems unwise to rely on this - with more load, of course the interval between gen 1 GCs may drop below the decay_time.

it's always a worry that when you change the existing perf behavior you will regress someone, thus the concern. 

also the idea of the fraction is we take some percentage from the current budget and some from the previous budget, but now with decay_factor we are simply discounting some of budget. shouldn't this be 

```
float prev_allocation_factor = (1.0 - allocation_fraction) * decay_factor;

new_allocation = (size_t)((1.0 - prev_allocation_factor) * new_allocation + 
                          prev_allocation_factor * previous_desired_allocation);
```
if we want to make the previous budget count for less based on the time factor?

**Comment by PeterSolMS** [comment]
> also the idea of the fraction is we take some percentage from the current budget and some from the previous budget, but now with decay_factor we are simply discounting some of budget. shouldn't this be

float prev_allocation_factor = (1.0 - allocation_fraction) * decay_factor;

new_allocation = (size_t)((1.0 - prev_allocation_factor) * new_allocation + 
                          prev_allocation_factor * previous_desired_allocation);

I think you are absolutely right. Will change the code accordingly.

**Review by Maoni0 — APPROVED** [review]
LGTM!

**Pull request merged**